src/DateTimeParser.jl

```diff
@@ -0,0 +1,512 @@
+                            tokens[i+1] in PERTAIN)
+                        # Jan of 01
+                        # In this case, 01 is clearly year
+                        try
+                            # Make a guess
+                            value = parse(Int, tokens[i+3])
+                            # Convert it here to become unambiguous
+                            push!(ymd, convertyear(value))
+                        end
+                        i += 4
+                    end
+                end
+            elseif haskey(AMPM, lowercase(tokens[i]))
+                # am/pm
+                value = AMPM[lowercase(tokens[i])]
+                if value == 2 && res["hour"] < 12
+                    res["hour"] += 12
+                elseif value == 1 && res["hour"] == 12
+                    res["hour"] = 0
+                end
+                i += 1
+            elseif haskey(res, "hour") && !haskey(res, "tzname") &&
+                    !haskey(res, "tzoffset") &&
+                    ismatch(r"^\w*$", tokens[i]) && !(lowercase(tokens[i]) in JUMP)
+                # Timezone name
+                res["tzname"] = tokens[i]
+                while i+2 <= len && tokens[i+1] == "/"
+                    res["tzname"] = string(res["tzname"], "/", tokens[i+2])
+                    i += 2
+                end
+                i += 1
+                # Check for something like GMT+3, or BRST+3. Notice
+                # that it doesn't mean "I am 3 hours after GMT", but
+                # "my time +3 is GMT". If found, we reverse the
+                # logic so that timezone parsing code will get it
+                # right.
+                if i <= len && tokens[i] in ("+", "-")
+                    tokens[i] = tokens[i] == "+" ? "-" : "+"
+                    if lowercase(res["tzname"]) in UTCZONE
+                        # With something like GMT+3, the timezone
+                        # is *not* GMT.
+                        delete!(res, "tzname")
+                    end
+                end
+            elseif haskey(res, "hour") && tokens[i] in ("+", "-")
+                # Numbered timzone
+                signal = tokens[i] == "+" ? 1 : -1
+                try
+                    i += 1
+                    tokenlength = length(tokens[i])
+                    if tokenlength == 4
+                        # -0300
+                        res["tzoffset"] = parse(Int, tokens[i][1:2])*3600+parse(Int, tokens[i][3:end])*60
+                    elseif i+1 <= len && tokens[i+1] == ":"
+                        # -03:00
+                        res["tzoffset"] = parse(Int, tokens[i])*3600+parse(Int, tokens[i+2])*60
+                        i += 2
+                    elseif tokenlength <= 2
+                        # -[0]3
+                        res["tzoffset"] = parse(Int, tokens[i])*3600
+                    else
+                        error("Faild to read timezone offset after +/-")
+                    end
```

Comment:
Why not separate the hour/minute extraction from parsing them? Should make the code cleaner.

```
hour, minute = tokens[i][1:2], tokens[i][3:end]
...
res["tzoffset"] = parse(Int, hour) * 3600 + parse(Int, minute) * 60
```